### DISCUSSION TITLE
Slow initial memory allocation

### DISCUSSION THREAD

**Discussion by otisdog8** [phil-opp/blog_os#844] [Questions]
I'm following the tutorial, and wanted to increase the size of the heap. This is the code for my allocator.rs:
``` rust
use x86_64::{structures::paging::{mapper::MapToError, FrameAllocator, Mapper, Page, PageTableFlags, Size4KiB, }, VirtAddr, };
use linked_list_allocator::LockedHeap;

#[global_allocator]
static ALLOCATOR: LockedHeap = LockedHeap::empty();

pub const HEAP_START: usize = 0x_4444_4444_0000;
pub const HEAP_SIZE: usize = 1 * 1024 * 1024; // 64 MiB - we want a heap this large so that it can contain the entire fs in ram, plus more

pub fn init_heap(mapper: &mut impl Mapper<Size4KiB>, frame_allocator: &mut impl FrameAllocator<Size4KiB>) -> Result<(), MapToError<Size4KiB>> {
    let page_range = {
        let heap_start = VirtAddr::new(HEAP_START as u64);
        let heap_end = heap_start + HEAP_SIZE - 1u64;
        let heap_start_page = Page::containing_address(heap_start);
        let heap_end_page = Page::containing_address(heap_end);
        Page::range_inclusive(heap_start_page, heap_end_page)
    };

    for page in page_range {
        let frame = frame_allocator
            .allocate_frame()
            .ok_or(MapToError::FrameAllocationFailed)?;
        let flags = PageTableFlags::PRESENT | PageTableFlags::WRITABLE;
        unsafe {
            mapper.map_to(page, frame, flags, frame_allocator)?.flush()
        };
    }

    unsafe {
        ALLOCATOR.lock().init(HEAP_START, HEAP_SIZE);
    }

    Ok(())
}
```
As of right now, I've increased the heap size to be 1 MiB. However, I would like to increase it more, and setting it where I want it (64 MiB) makes the OS take about two minutes to boot. I believe the main cause of the slowdown is the `for page in page_range`, but I'm not sure if that's the actual cause.

How can I shorten boot time while still allocating large quantities of memory?

**Comment by bjorn3** [comment]
Two questions: Are you compiling in release mode? And are you using hardware assisted virtualization? (eg KVM for qemu)

**Comment by otisdog8** [comment]
Not using KVM and not in release mode. Switching to release mode makes it significantly faster (10 seconds or so). Switching to release mode + kvm (with -accel kvm) makes it even faster (2 seconds or so), but significantly worsens VGA graphics performance (specifically the 640 x 480 x 16 color mode).

**Comment by toothbrush7777777** [comment]
@otisdog8 You could try allocating large (2 MiB) pages ([Size2MiB](https://docs.rs/x86_64/latest/x86_64/structures/paging/page/enum.Size2MiB.html)). That should improve performance significantly.
